apps/state/lib/state/connecting_stops.ex

```diff
@@ -0,0 +1,192 @@
+      |> Stream.map(fn stop -> [stop | connecting_stops(stop, overrides)] end)
+      |> Stream.flat_map(&rotations/1)
+      |> group_and_merge_by_head()
+      |> Stream.map(fn {stop, stops} -> {stop, reject_redundant_connections(stop, stops)} end)
+      |> Enum.map(fn {stop, stops} -> {stop.id, Enum.map(stops, & &1.id)} end)
+
+    :ets.delete_all_objects(@table)
+    :ets.insert(@table, items)
+    size = length(items)
+
+    _ =
+      Logger.info(fn ->
+        "Update #{__MODULE__} #{inspect(self())}: #{size} items"
+      end)
+
+    publish({:new_state, __MODULE__}, size)
+    state
+  end
+
+  # Find possible connecting stops for a given stop, taking overrides into account.
+  @spec connecting_stops(Stop.t(), overrides) :: [Stop.t()]
+  defp connecting_stops(stop, overrides) do
+    standalone_stops_around(stop)
+    |> MapSet.new()
+    |> MapSet.union(MapSet.new(override_stops(overrides, stop, :add)))
+    |> MapSet.difference(MapSet.new(override_stops(overrides, stop, :remove)))
+    |> MapSet.to_list()
+  end
+
+  # Find "standalone" stops close to a given stop's entrances, if it has entrances, otherwise
+  # close to the stop itself.
+  @spec standalone_stops_around(Stop.t()) :: [Stop.t()]
+  defp standalone_stops_around(%{id: stop_id, latitude: latitude, longitude: longitude}) do
+    case entrances(stop_id) do
+      [] ->
+        State.Stop.around(latitude, longitude, @radius) |> Enum.filter(&Stop.standalone?/1)
+
+      entrances ->
+        entrances |> Stream.flat_map(&standalone_stops_around/1) |> Enum.uniq()
+    end
+  end
+
+  @spec entrances(Stop.id()) :: [Stop.t()]
+  defp entrances(parent_station) do
+    State.Stop.Cache.match(%{location_type: 2, parent_station: parent_station}, :parent_station)
+  end
+
+  # Get stops that should be added/removed from a given parent station's connections.
+  @spec override_stops(overrides, Stop.t(), :add | :remove) :: [Stop.t()]
+  defp override_stops(overrides, %{id: stop_id}, kind) when kind in [:add, :remove] do
+    overrides |> Map.get(stop_id, %{}) |> Map.get(kind, []) |> State.Stop.by_ids()
+  end
+
+  # Return all rotations of a non-empty list.
+  # Example: [1, 2, 3] => [[1, 2, 3], [2, 3, 1], [3, 1, 2]]
+  @spec rotations([...]) :: [[...]]
+  defp rotations([item]), do: [[item]]
+
+  defp rotations([_ | _] = items) do
+    0..(length(items) - 1)
+    |> Enum.map(fn count ->
+      {before, [first | rest]} = Enum.split(items, count)
+      [first | before ++ rest]
+    end)
+  end
+
+  # Given an enumerable of lists, split them into heads and tails, group tails that share a head,
+  # then flatten and uniq the tails for each head.
+  # Example: [[1, 2, 3], [1, 3, 4], [2, 3, 4]] => [{1, [2, 3, 4]}, {2, [3, 4]}]
+  @spec group_and_merge_by_head(Enumerable.t()) :: Enumerable.t()
+  defp group_and_merge_by_head(combinations) do
+    combinations
+    |> Enum.group_by(&hd/1, &tl/1)
+    |> Stream.map(fn {head, tail} -> {head, tail |> List.flatten() |> Enum.uniq()} end)
```

Comment:
Might be more efficient to merge the tails with `Enum.reduce(tail, MapSet.new(), fn t, set -> MapSet.union(set, MapSet.new(t)) end)` or `tail |> Enum.map(&MapSet.new/1) |> Enum.reduce(&MapSet.union/1)`

Reply:
I checked and this does appear to shave ~15ms off the overall 125ms update process (on current GTFS) — but given how fast it is already and how infrequently it will be updating, I think that difference doesn't really matter, and for me it's outweighed by the current version being a bit easier to parse (er... with your eyeballs).

```elixir
Stream.map(fn {head, tails} -> {head, tails |> List.flatten() |> Enum.uniq()} end)
```

versus

```elixir
Stream.map(fn {head, tails} ->
  {head, tails |> Enum.map(&MapSet.new/1) |> Enum.reduce(&MapSet.union/2) |> MapSet.to_list()}
end)
```

I did just realize I mis-named this variable and it should have been `tails` (plural), so will fix that!

Reply:
Does it need to be a list at the end?

Reply:
I don't suppose it needs to be, though it would still require this to be 3 lines. If you feel strongly about the readability in the other direction, I don't mind changing it.

Reply:
not sure what "three lines" means, but I don't think this is super-important.

Reply:
Ah, like, it would have to be formatted as 3 lines, for max line length reasons. Anyway 👍